### PR TITLE
fix(executor): clean up process trees on command timeout

### LINKED ISSUE #350: fix(executor): executor timeout may leave process tree uncleaned, permanently blocking a command worker
## Problem

The executor's timeout handling relies solely on `exec.CommandContext`'s default kill behavior, and `cmd.Wait()` is called without setting `WaitDelay`.

Locations:
- `pkg/executor/executor.go:212` — `exec.CommandContext`. On ctx cancellation it sends SIGKILL to the parent process only via `cmd.Process.Kill()`. It is not a process-group kill.
- `pkg/executor/executor.go:338` — `cmd.Wait()` with no `WaitDelay` set (`WaitDelay=0`, the default).
- `pkg/executor/session_unix.go` — sets `Setsid=true`, but only when `opts.PIDHook != nil`, and it never group-kills on timeout.
- `pkg/executor/session_windows.go` — `enableSessionLeader` is a no-op. No Job Object.

Because `cmd.Stdout`/`cmd.Stderr` are in-memory writers (`chunkWriter` or `bytes.Buffer`) rather than `*os.File`, Go sets up an internal `os.Pipe` and a copy goroutine, and `cmd.Wait()` blocks until that goroutine sees EOF. If only the parent is killed while a grandchild that inherited the pipe write-end stays alive, EOF never arrives and `cmd.Wait()` blocks forever. With `WaitDelay=0`, Go does not force-close the pipes after the kill.

Command execution runs inside the worker pool, so if this state recurs, workers are never returned and command-execution availability degrades. The default worker count is 20 (`DefaultPoolMaxWorkers`, `pkg/config/config.go:33`); 10 is the `maxWorkers <= 0` fallback in `NewPool` (`internal/pool/pool.go:30`).

## Impact

- A timed-out command can permanently occupy a worker.
- If it recurs, agent command execution can stall entirely.
- On Windows, the absence of a Job Object means child processes can be orphaned.
- Reproduced via `ExecWithHook` (the Start/Wait path), but the `cw == nil && pidHook == nil` `CombinedOutput` path (`executor.go:313-314`) is equally vulnerable. That means `RunWithTimeout`/`Exec`/`RunWithInput` can block the same way.

## Reproduction

Reproduced on darwin/unix.

```
/bin/sh -c "sleep 600 & echo $! > child.pid; wait"
```

Running the above through the executor's streaming/pid-hook path with a 200ms timeout: the `sh` parent is killed, but the backgrounded `sleep 600` keeps holding the stdout/stderr pipe fd it inherited, so the executor does not return even after the 200ms timeout. `cmd.Wait()` only unblocks once the leaked child PID is SIGKILLed directly.

## Fix direction

- Add a shared timeout/cancel abstraction.
- Unix: create a process group (`Setpgid`), then group-kill (`kill(-pgid)`) on timeout.
- Windows: Job Object + `JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE`.
- Set `cmd.WaitDelay`.
- Apply to both the Start/Wait path and the `CombinedOutput` path.
- Add tests for the timeout/kill/Wait paths.

## Acceptance criteria

- On timeout, child and grandchild processes are terminated.
- `cmd.Wait()` returns even when a child is holding a pipe open.
- Regression or platform-specific tests added for Unix and Windows.

### PR DESCRIPTION
**Pull request by jisung-02**
## Summary

When a command times out, the executor only killed the immediate child process. If a descendant (e.g. a backgrounded process) inherited and kept holding the stdout/stderr pipe open, `cmd.Wait()` blocked forever waiting for EOF, permanently occupying a worker in the pool.
On Windows there was no Job Object, so orphaned children could survive indefinitely.

Reproduction: `/bin/sh -c "sleep 600 & echo $! > pidfile; wait"` run through the executor with a short timeout. The `sh` parent gets killed, but the backgrounded `sleep` keeps the inherited pipe fd open, so `Wait()` never returns.

## Changes

- `pkg/executor/executor.go`: `runCommand` now always uses `Start`/`Wait` (previously `CombinedOutput` for the plain case, which had the same leaked-pipe exposure). Sets `cmd.WaitDelay` and `cmd.Cancel` so a timeout forces the process tree closed and unblocks `Wait()` even when a descendant is holding the pipe.
- `pkg/executor/executor.go`: the process tree is now torn down on every exit path via an unconditional deferred `cleanup.cancel(cmd)` after `Start`, not only on the timeout/`ErrWaitDelay` paths. A plain non-zero exit that leaves a descendant holding the pipe triggers neither `Cancel` (timeout-only) nor the `ErrWaitDelay` branch, so that descendant used to leak; `WaitDelay` still unblocks `Wait()`, so this was a leaked descendant rather than worker starvation. `cancel` is idempotent (returns `ErrProcessDone` once the tree is gone).
- `pkg/executor/process_tree_unix.go` (new): puts the child in its own session or process group (`Setsid`/`Setpgid`) and kills the whole group with `syscall.Kill(-pgid, SIGKILL)` on cancel. The process group is captured in `afterStart` while the leader is still alive, because `getpgid(pid)` fails with `ESRCH` once `Wait` reaps the leader; `-pgid` stays a valid group-kill target as long as a descendant keeps the group alive. It is recorded only when the child leads its own group (`getpgid(pid) == pid`), so `-pgid` can never hit an unrelated group. When a PID hook is registered (sudo-tracked exec/deploy shell path), the child stays a session leader so PAM/sudo session resolution keeps working, matching the previous `enableSessionLeader` behavior.
- `pkg/executor/process_tree_windows.go` (new): assigns the child to a Job Object with `JOB_OBJECT_LIMIT_KILL_ON_JOB_CLOSE`, plus a `CreateToolhelp32Snapshot`-based process-tree walk as a fallback for descendants that fork before the Job Object assignment completes. Each cleanup method confines its mutex to a small `defer`-released helper so the slow Win32 kills run lock-free.
- `pkg/executor/process_tree.go` (new): a build-tag-neutral `commandCleaner` interface (`afterStart`/`cancel`/`close`). Each platform file asserts `var _ commandCleaner = (*commandCleanup)(nil)`, so a method-signature drift between the Unix and Windows cleanup types fails that platform's build early instead of only surfacing when the opposite platform's CI runs.
- `pkg/executor/executor.go`: `exec.ErrWaitDelay` from `Wait()` means the command's own process already exited successfully and only a lingering descendant kept the pipe open past the delay; that path is reported as success (exit 0) instead of exit 1 with Go's internal `"exec: WaitDelay expired..."` string leaking into the result. The `cmd.Start` failure path is routed through the shared `finish` helper for consistency with the other return paths.
- Removed `session_unix.go`/`session_windows.go`; the session-leader logic moved into `configureProcessTreeCleanup`.
- `pkg/executor/executor.go`: incidental cleanup of the existing streaming path, unrelated to the timeout fix—moved `runeSafeCut` below the `chunkWriter` cluster it belongs to, and switched the flusher goroutine to `sync.WaitGroup.Go` (Go 1.25). No behavior change.

## Tests

- `pkg/executor/process_tree_unix_test.go` (new): white-box unit tests for the Unix cleanup contract. `configureProcessTreeCleanup` is covered by a flag matrix (the `Setpgid`/`Setsid` branches, a preexisting-`Setsid` command that must not get `Setpgid` forced on, and preservation of an existing `Credential` from `utils.Demote`); `cancel` is covered for the nil-process case (`os.ErrProcessDone`) and a direct process-group kill of a backgrounded grandchild.
- `pkg/executor/process_tree_unix_test.go` also holds two black-box regression tests (moved/added here so process-tree tests sit next to the code they exercise): the timeout case (`TestExecutor_TimeoutCleansProcessTreeWhenChildKeepsPipeOpen`, both the plain and PID-hook paths) asserts exit 124 promptly with the child gone; and the non-zero-exit case (`TestExecutor_CleansDescendantWhenCommandExitsNonZero`) covers the path with no timeout where `Wait` returns an `ExitError` rather than `ErrWaitDelay`, asserting the backgrounded descendant is still reaped by the unconditional cancel.
- The timeout test's liveness check (`waitForExecutorTimeoutChildExit`) failed on every Linux CI job while passing on macOS/Windows. GitHub Actions runs each Linux distro job via `docker exec` into an already-running container, so the test process is never PID 1; when the timed-out shell is killed, its `sleep` grandchild gets reparented but nothing in that PID namespace reaps it, so it becomes a zombie. A zombie still answers `kill(pid, 0)` successfully (not `ESRCH`), which is what the check used as "process is gone", so it spun for the full window and failed even though the executor had already SIGKILLed the process group correctly. Fixed by also accepting a zombie (`/proc/<pid>/stat` state `Z`) as terminated on Linux. Reproduced and verified locally with `docker run -d <image> sleep infinity` + `docker exec`, matching GitHub Actions' container-job model.
- `pkg/executor/process_tree_windows_test.go`: white-box tests for the Windows cleanup path. Covers the Job Object assignment path (single process, and a multi-level `cmd.exe -> ping` tree), the `CreateToolhelp32Snapshot` PID-tree-walk fallback when Job assignment is skipped (single process; the multi-level tree is exercised on the Job path above), the `afterStart` cancel-race guard (cancel arriving before `afterStart` records the pid/handle), and a black-box exit-124 timeout test mirroring the Unix one. These are compiled with `GOOS=windows go vet` but are not executed on the dev host; they need a Windows runner.
- `pkg/executor/executor_unix_test.go`: now holds only the environment-isolation test (`TestExecutor_DoesNotInheritProcessEnv`); the process-tree tests moved to `process_tree_unix_test.go`.

## Test plan

- [x] `go build ./pkg/executor/`
- [x] `go vet ./pkg/executor/`
- [x] `GOOS=windows go build ./pkg/executor/`
- [x] `GOOS=windows go vet ./pkg/executor/` (Windows-tagged tests compile; not executed on darwin)
- [x] `gofmt -l pkg/executor/*.go` (clean)
- [x] `go test ./pkg/executor/... -p 1 -count=1` (all non-Windows tests pass, including the new white-box unit tests and both timeout / non-zero-exit cleanup regression tests)
- [x] Reproduced the Linux CI failure locally via `docker run -d golang:1.25 sleep infinity` + `docker exec` and confirmed the zombie-check fix resolves it in that same container
- [ ] Windows tests in `process_tree_windows_test.go` still need to run on a Windows runner (vet-checked only so far)

## Notes

Windows Job Object assignment happens after `cmd.Start()`, so a descendant that forks in that window can escape the job; the toolhelp tree-walk fallback exists specifically to catch that case and is not fully redundant with `KILL_ON_JOB_CLOSE`.

Cleanup is a hard kill with no SIGTERM grace period: `SIGKILL` to the process group on Unix, `TerminateProcess` + `KILL_ON_JOB_CLOSE` on Windows. This is deliberate for a timeout-reclaim path—a descendant holding the pipe open may ignore SIGTERM anyway—but it means a child's own cleanup (temp files, etc.) is skipped, and GNU `timeout`'s default SIGTERM-then-SIGKILL is intentionally not replicated.

On Unix, `cancel` group-kills (`kill(-pgid)`) only for a child that led its own group at start time (recorded in `afterStart`), so `-pgid` can never target an unrelated group. The unconditional post-`Wait` cancel accepts a narrow PID-reuse TOCTOU window: after the whole tree has exited and the group is empty, the pgid becomes reusable, so a rapid reuse into a new group leader before the kill fires could in theory be hit. This is the same risk class the timeout path already carried, and is far narrower than the original bug (permanent worker starvation).

Closes #350

### CODE REVIEW

File: pkg/executor/process_tree_windows.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+	c.mu.Lock()
+	c.pid = pid
+	c.mu.Unlock()
+
+	handle, err := windows.OpenProcess(
+		windows.PROCESS_SET_QUOTA|windows.PROCESS_TERMINATE,
+		false,
+		pid,
+	)
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/process_tree_windows.go:64]: Fixed. Retained the process handle opened during job assignment (`commandCleanup.handle`) instead of closing it right away, and cancel/close now terminate/close via that handle first. A handle keeps referring to the same process object regardless of PID reuse, so the retained handle is now the primary, reuse-proof kill path; the PID-based toolhelp tree walk stays only as the fallback for descendants and for the case where no handle was ever obtained.

Also hit a related bug while validating this on the actual Windows CI runner: once the Job Object's KILL_ON_JOB_CLOSE already terminated the process, a subsequent TerminateProcess on the retained handle (or on a PID re-opened in the tree walk) returns ERROR_ACCESS_DENIED rather than an "already exited" code. Added `isWindowsTerminateGone` to treat that as success at both TerminateProcess call sites.

File: pkg/executor/process_tree_windows.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	if err != nil {
+		return err
+	}
+	defer windows.CloseHandle(handle)
+
+	if err := windows.AssignProcessToJobObject(c.job, handle); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/process_tree_windows.go]: Fixed. `tryAssignToJob` now returns a bool instead of propagating an error: if OpenProcess or AssignProcessToJobObject fails (e.g. under a restrictive job policy), `assigned` just stays false and the command still runs — cancel falls back to the PID-based toolhelp tree walk (and the retained handle when we got one) instead of the command failing outright.

File: pkg/executor/process_tree_windows.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	closed   bool
+}
+
+func configureProcessTreeCleanup(_ *exec.Cmd, _ bool) (*commandCleanup, error) {
+	job, err := windows.CreateJobObject(nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &commandCleanup{job: job}, nil
+}
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply pkg/executor/process_tree_windows.go:36]: Fixed. Added `process_tree_windows_test.go` with two tests covering `commandCleanup` directly: one spawns a real process, confirms job assignment succeeds, and verifies cancel terminates it; the other skips assignment to exercise the toolhelp-based PID tree walk on its own. Both run on the `windows-amd64` CI job (this sandbox has no Windows runtime, so I validated by pushing and reading the actual CI logs — that loop is also what surfaced the ERROR_ACCESS_DENIED bug fixed alongside this).

This isn't the same shape as the Unix test (a real inherited-pipe leak through a backgrounded grandchild) — that would need a self-exec Go helper to reproduce reliably on Windows, which felt like a larger, separate effort. Left as a known follow-up rather than guessing at an unverified implementation.

**Review by geunwoonoh — DISMISSED** [review]
## Review: #353 - fix(executor): clean up process trees on command timeout

**Stack**: Go (golang)
**Local verification**: `go build` / `go vet` / `gofmt` / `GOOS=windows build+vet` / `go test ./pkg/executor/... -p 1` all pass ✅

---

### Summary

This PR improves timeout handling so that instead of killing only the immediate child, the executor tears down the entire process tree, including grandchildren that keep an inherited stdout/stderr pipe fd open. Key mechanisms:

- `runCommand` now always uses `Start`/`Wait` and sets `cmd.WaitDelay` (2s) + `cmd.Cancel`, so `Wait()` no longer blocks forever when a descendant holds the pipe open.
- Unix: creates a process group via `Setsid`/`Setpgid` and group-kills with `kill(-pid, SIGKILL)`, gated by a `getpgid(pid) == pid` leader check.
- Windows: Job Object (`KILL_ON_JOB_CLOSE`) plus a Toolhelp snapshot process-tree walk fallback.
- `enableSessionLeader` logic was folded into `configureProcessTreeCleanup`, and `session_{unix,windows}.go` were removed.

Overall this is a very solid, high-quality PR. All three Copilot Windows findings (PID reuse, aborting the command on Job-assignment failure, missing Windows test) have already been addressed. Beyond those, I found only one substantive gap and a few minor suggestions.

---

### Critical Issues (0)

None.

---

### Warnings (1)

**1. `pkg/executor/executor.go:362-367`** — a descendant is not cleaned up in the "non-zero exit + descendant holding the pipe" combination.

```go
err = cmd.Wait()
if errors.Is(err, exec.ErrWaitDelay) {
    cleanup.cancel(cmd) // kill the descendant tree
}
```

I checked the `os/exec` `Wait()` source in the Go 1.25.11 toolchain directly (error precedence):

```go
state, err := c.Process.Wait()
if err == nil && !state.Success() {
    err = &ExitError{ProcessState: state}   // <- for a non-zero exit, err is fixed here
}
...
if err == nil && watch.err != nil { err = watch.err }        // WaitDelay/ctx err only applied when err == nil
if goroutineErr := c.awaitGoroutines(timer); err == nil {    // ErrWaitDelay only applied when err == nil
    err = goroutineErr
}
```

So when **the command itself exits non-zero while a descendant holds the pipe open**, `Wait()` returns an `ExitError`, not `ErrWaitDelay`. As a result the `errors.Is(err, exec.ErrWaitDelay)` branch above is not taken, `cleanup.cancel()` is never called, and **the descendant is left as an orphan**.

- Repro: `sh -c "sleep 600 & exit 3"` (without a timeout).
- That said, this is still much better than the original bug (permanent worker starvation). `WaitDelay` closes the pipe, so `Wait()` returns after 2s and **there is no worker starvation**; only a leaked descendant remains. The timeout path and the exit-0 path both clean up the tree correctly.
- I raise it because it is a case the PR's goal ("clean up process trees") and the Notes section don't cover.

Suggested options (pick one):
- Run an idempotent group-kill unconditionally after `Wait()` via `defer cleanup.cancel(cmd)`. The Unix `cancel` has a `getpgid==pid` guard and `ESRCH -> ProcessDone` handling, so it is harmless when the leader is already dead (there is a very narrow PID-reuse TOCTOU window).
- Or at minimum document this in the Notes: "a descendant holding the pipe open is not reaped when the command exits non-zero."

---

### Suggestions (3)

**1. `pkg/executor/process_tree_unix.go:14` / `pkg/executor/process_tree_windows.go`** — the two `commandCleanup` types have no compile-time contract enforcement.

The comment acknowledges this ("unenforced across build tags, so keep the two in sync"). This is a common tradeoff with Go build tags, but if the `afterStart`/`cancel`/`close` signatures drift apart across platforms, it won't be caught until the opposite platform's CI runs. Declaring a shared interface plus a compile-time assertion in a build-tag-neutral file would catch drift earlier. (Note: Unix uses a value receiver and Windows a pointer receiver, so this would require changing the return type to the interface — worth weighing cost vs. benefit.)

**2. `pkg/executor/executor.go:343-348`** — the `cmd.Start()` failure path bypasses `finish`.

```go
if err := cmd.Start(); err != nil {
    if cw != nil {
        return nil, err        // does not go through finish() (no cw.close())
    }
    return buf.Bytes(), err
}
```

The flusher goroutine hasn't started yet, so this is functionally harmless, but every other return path uses `finish(err)`. Unifying this to `return finish(err)` would read more consistently (the `cw == nil` case is equivalent via `buf.Bytes()`).

**3. `pkg/executor/executor.go:349-355`** — double `cancel` call on the afterStart error path (Windows race only).

If Windows `afterStart` internally already called `cancel` and then returns non-nil, `runCommand` calls `cancel` again. It's idempotent so there's no behavioral issue, but a one-line comment ("afterStart may have already canceled, but cancel is idempotent") would avoid future confusion. Unix `afterStart` always returns nil, so it's unaffected.

---

### Go conventions / duplication & dead code / performance (additional review points)

**Go conventions** — idiomatic overall.
- Naming (`configureProcessTreeCleanup`, `terminateWindowsProcessTree`, `takeForCancel`, etc.), error wrapping (`errors.Is`), build-tag separation, and confining lock scope with `defer` are all good.
- `sync.WaitGroup.Go` (new in Go 1.25) is used; go.mod is `go 1.25.11`, so this is fine.

**Duplication / dead code** — none.
- No lingering references to the removed `enableSessionLeader` / `session_{unix,windows}.go` (verified via grep).
- `runeSafeCut` was only relocated below the `chunkWriter` cluster; behavior unchanged.
- `TestExecutor_TimeoutCleansProcessTreeWhenChildKeepsPipeOpen` exists in both the unix and windows test files under the same name, but they are mutually exclusive via build tags, so there is no real duplicate compilation (the normal pattern for parallel per-platform tests).

**Performance** — no regression, and if anything more defensive.
- The plain path changed from `CombinedOutput` to `Start`/`Wait`, but `CombinedOutput` internally does the same `Start`+`Wait`+buffer, so overhead is identical.
- Windows adds one `CreateJobObject` syscall per command — negligible, and justified by the tree-cleanup guarantee.
- `WaitDelay = 2s` only introduces delay on the leaked-pipe path; no effect on the normal path.
- Splitting the slow Win32 kills out of the `mu` critical section (lock-free) via the `take*` helpers is a nice design choice.

---

### Highlights

- Group-kills with `-pid` only after confirming `getpgid(pid) == pid`, preventing accidental kills of an unrelated group. Nicely defensive.
- Normalizing `ErrWaitDelay` to exit 0 prevents Go's internal string (`"exec: WaitDelay expired..."`) from leaking into the result. (`pkg/executor/executor.go:288-291`)
- Handling the case where an orphaned grandchild in a CI container is never reaped and stays a zombie (`kill(pid,0)` does not return ESRCH) by checking `/proc/<pid>/stat` state `Z` — great catch, reproduced against the actual GitHub Actions docker-exec model.
- Addressing all three Copilot findings (PID reuse -> retained handle, Job-assignment failure -> non-fatal fallback, adding Windows tests) while verifying against CI logs, and incidentally catching and fixing the `ERROR_ACCESS_DENIED` bug in the process.

---

**Verdict**: APPROVE (no merge blockers)

Warning 1 is a clear improvement over the original bug (worker starvation); all that remains is an orphan leak in a narrow edge case. It's not merge-blocking and can be handled in a follow-up commit or a Notes addition.

**Checklist**:
- [x] Follows Go conventions
- [x] Tests included and passing (Unix white/black-box + timeout regression; Windows is vet-only, runs in CI)
- [x] No security vulnerabilities (no hardcoded secrets/injection; the SIGKILL group-kill is scoped by the leader guard)
- [x] Docs/Notes updated (only Warning 1's case is worth adding)
- [ ] The Windows inherited-pipe black-box test is left as a known follow-up by the author

**Comment by jisung-02** [comment]
Thanks for the review. Addressed all of it in 3ceff53.

Warning 1 (descendant leaked on a non-zero exit): `runCommand` now tears the tree down on every exit path via an unconditional deferred `cancel`, not just on the timeout/`ErrWaitDelay` paths. The literal "defer cancel" alone wasn't enough on Unix: once `Wait` reaps the leader, `getpgid(pid)` returns `ESRCH`, so the cancel-time guard fell back to a positive-PID kill and missed the group. So `cancel` no longer re-derives the group at cancel time — `afterStart` captures the pgid while the leader is still alive, and `-pgid` stays a valid group-kill target while a descendant keeps the group open. Added a regression test (`TestExecutor_CleansDescendantWhenCommandExitsNonZero`).

Suggestion 1: added a build-tag-neutral `commandCleaner` interface with per-platform `var _ commandCleaner = (*commandCleanup)(nil)` assertions, so a signature drift fails that platform's build early. Kept the concrete `*commandCleanup` return type since the Windows tests read concrete fields.

Suggestion 2: routed the `cmd.Start` failure path through `finish()`.

Suggestion 3: added the afterStart/cancel-race comment on the afterStart error branch.

PR body updated to match.

**Review by geunwoonoh — APPROVED** [review]
